bpo-46218: Change long_pow() to sliding window algorithm

tim-one opened the pull request:
For large exponents in long_pow(), use the sliding window algorithm instead.

Also boost the window size from 5 bits to 6, which should yield a modest but significant speedup for long exponents. The precomputed table remains the same size, though, because the sliding window algorithm only stores results for odd exponents.

long_pow() no longer requires that the number of bits in a CPython long digit be a multiple of 5. It no longer cares at all what the digit width is.

<!--
Thanks for your contribution!
Please read this comment in its entirety. It's quite important.

# Pull Request title

It should be in the following format:

```
bpo-NNNN: Summary of the changes made
```

Where: bpo-NNNN refers to the issue number in the https://bugs.python.org.

Most PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.

# Backport Pull Request title

If this is a backport PR (PR made against branches other than `main`),
please ensure that the PR title is in the following format:

```
[X.Y] <title from the original PR> (GH-NNNN)
```

Where: [X.Y] is the branch name, e.g. [3.6].

GH-NNNN refers to the PR number from `main`.

-->


<!-- issue-number: [bpo-46218](https://bugs.python.org/issue46218) -->
https://bugs.python.org/issue46218
<!-- /issue-number -->

tim-one committed:
bpo-46218: Change long_pow() to sliding window algorithm

For large exponents in long_pow(), use the sliding window algorithm instead.

Also boost the window size from 5 bits to 6, which should yield a modest but significant speedup for long exponents. The precomputed table remains the same size, though, because the sliding window algorithm only stores results for odd exponents.

long_pow() no longer requires that the number of bits in a CPython long digit be a multiple of 5. It no longer cares at all what the digit width is.

tim-one committed:
Fix spelling error.

rhettinger commented on Lib/test/test_pow.py:
```diff
@@ -93,6 +93,28 @@ def test_other(self):
+        #  a**(b1 + b2) == a**b1 * a**b2
+        #  a**(b1 * b2) == (a**b1)**b2
+        prime = 1000000000039 # for speed, relatively small prime modulus
+        for i in range(10):
```
Consider modernizing these tests.  Put _a_, _b1_, and _b2_ in a subtest.  Use assertEqual for the _got_ compares.  If there is a failure, the subtest logic with print inputs and the assertEqual logic will print the outputs.

tim-one replied:
These are randomized tests of large ints. Seeing the computed values on failure is useless: you need to know `a`, `b1`, and `b2` to reproduce the failure. I also wanted to force hex output of those, to avoid quadratic-time base conversion for failure output.

rhettinger replied:
The hex output is the key then.   Seeing *a*, *b1*, *b2* is what subtest() does for you.

tim-one replied:
OK - I didn't know the `subtest` context manager existed - I avoid unittest whenever conveniently possible :wink:. Instead I fiddled what I wrote to fail on purpose until I could use the failure output to easily reconstruct the failing case. What's there now works fine for that purpose, and it was easy to write it in a way that did so. Is there a point to using the latest elaborate framework feature to do instead what can be easily & obviously done without it? If you were to say you have a hard time _understanding_ the code that's there, then I'd struggle to use the subtest gimmick instead. Else I'd rather not bother.

rhettinger commented on Lib/test/test_pow.py:
```diff
@@ -93,6 +93,28 @@ def test_other(self):
+            bpower = random.randrange(1000, 5000)
+            b = random.randrange(1 << (bpower - 1), 1 << bpower)
+            b1 = random.randrange(1, b)
+            b2 = b - b1
```
Why `b - b1` instead of `random.randrange(1, b)`?

tim-one replied:
No compelling reason; just wanted to guarantee that it would never fall into using two small exponents (if `b1` happens to  be smallish, `b2` cannot be; & vice versa).

rhettinger commented on Objects/longobject.c:
```diff
@@ -74,12 +74,15 @@ maybe_small_long(PyLongObject *v)
+/* For exponentiation, use the binary left-to-right algorithm unless the
+ ^ exponent contains more than HUGE_EXP_CUTOFF bits.  In that case, do
+ * (no more than) EXP_WINDOW_SIZE bits at a time.  The potential drawback is
+ * hat a table of 2**(EXP_WINDOW_SIZE - 1) intermediate results is
```
`s/hat/that`

tim-one replied:
That change got checked in before I saw this 😃

rhettinger commented on Objects/longobject.c:
```diff
@@ -4366,23 +4371,65 @@ long_pow(PyObject *v, PyObject *w, PyObject *x)
-        for (i = 1; i < 32; ++i)
-            MULT(table[i-1], a, table[i]);
+        /* Left-to-right k-ary sliding window exponentiation
+         * (HAC Algorithm 14.85)
```
Spell-out "Handbook of Applied Cryptography".  Perhaps add a link: https://cacr.uwaterloo.ca/hac/about/chap14.pdf

tim-one replied:
There's already a link in the `x_mul()` comments, and yet another just above in the "left-to-right binary exponentiation" block, but I don't think they help. It's impossible not find the book at once via Google, but links can change. I'll spell out the acronym, though.

rhettinger commented on Objects/longobject.c:
```diff
@@ -4366,23 +4371,65 @@ long_pow(PyObject *v, PyObject *w, PyObject *x)
+         * current number of trailing zeroes `ntz`.
+         */
+        int pending = 0, blen = 0, ntz = 0;
+#define ABSORB_PENDING  do { \
```
Consider making this a static inline function that passes in the addresses of the variables.

tim-one replied:
Actively don't want to do that. This block is expanded exactly twice, will never be used anywhere else, and doesn't correspond to a coherent abstraction. It's just doing exactly what needs to be done in the two places it's expanded, and is a macro solely to minimize the chance of the two code blocks getting out of synch. Indeed, I'd rather type it out in full twice than try to pretend it's a coherent "function", full of noisy `*pending` and `*ntz`  and `*blen` obfuscations.

tim-one replied:
BTW, note too that `ABSORB_PENDING` contains instances of the `MULT` macro, which in turn contains use of the `REDUCE` macro, and both of the latter contain `goto Error;` statements. Those cannot be turned into functions, or used by other functions, without significant surgery to rework them (and also correspondingly change _all_ the code sites that use them).

More positively, I actively like having the body of `ABSORB_PENDING` so physically close to the small loop nest that uses that macro. They're exceedingly tightly coupled in what they do, and it would be a major PITA to have to scroll the editor multiple pages (`long_pow()` is a long function!) to flip between them.

tim-one committed:
Address a review comment.

sweeneyde commented on Objects/longobject.c:
```diff
@@ -4332,7 +4337,7 @@ long_pow(PyObject *v, PyObject *w, PyObject *x)
         /* else bi is 0, and z==1 is correct */
     }
-    else if (i <= FIVEARY_CUTOFF) {
+    else if (i * PyLong_SHIFT <= HUGE_EXP_CUTOFF) {
```
Can this multiplication overflow?

tim-one replied:
Not easily - but yes :wink: Good catch!

tim-one committed:
Rework tests to avoid potential multiply overflow noted in review.
Good catch! The code is clearer the new way too.

tim-one committed:
Simplify the main loop logic considerably by moving the
trailing zero logic entirely into ABSORB_PENDING. These
native int bit manipulations are dirt cheap in
comparison to the bigint squaring needed for each
exponent bit.

And boost the size of exponents tested to (probabilistically)
stress a greater mix of exponent bit patterns.

tim-one committed:
Cut window size back from 6 to 5 bits. Even for an exponent with
a million straight 1 bits, the timing difference seems insignicant.
So I think it better to cut the table size in half, to cut the
precomputation overhead time in half for "saner" (smaller
exponent) cases.

tim-one committed:
Reduce the threshold for switching to sliding window pow.

Since the dynamic table of small powers needed is half the
size now, the overhead of trying to use the k-ary method has
been cut accordingly, which allows it to pay off at smaller
exponent bit lengths.

sweeneyde commented on Objects/longobject.c:
```diff
@@ -4366,23 +4389,59 @@ long_pow(PyObject *v, PyObject *w, PyObject *x)
-        for (i = 1; i < 32; ++i)
-            MULT(table[i-1], a, table[i]);
+        /* Left-to-right k-ary sliding window exponentiation
+         * (Handbookd of Applied Cryptography (HAC) Algorithm 14.85)
```
Handbookd --> Handbook

tim-one committed:
Repair spelling error in comment, pointed out in review.

arhadthedev commented on Objects/longobject.c:
```diff
@@ -4172,14 +4194,15 @@ long_pow(PyObject *v, PyObject *w, PyObject *x)
      */
-    PyLongObject *table[32] = {0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,
-                               0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0};
+    PyLongObject *table[EXP_TABLE_LEN] = {0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0};
```
Can we use a shorthand? If an array is partially initialized, elements that are not initialized receive the value zero.

```suggestion
    PyLongObject *table[EXP_TABLE_LEN] = {0};
```

tim-one replied:
Ya, I've waffled on that. At heart, I want to keep it annoying to change `EXP_TABLE_LEN`, to force people to really think about possible consequences. But you're right, the code now just looks weird. Thanks!

tim-one committed:
Update Objects/longobject.c

Use shortcut initializer for the k-ary table.

Co-authored-by: Oleg Iarygin <dralife@yandex.ru>
